pagerduty/resource_pagerduty_user_contact_method.go

```diff
@@ -0,0 +1,173 @@
+				Computed: true,
+			},
+
+			"label": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+
+			"address": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+
+			"contact_method_id": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func buildUserContactMethodStruct(d *schema.ResourceData) *pagerduty.ContactMethod {
+	contactMethod := &pagerduty.ContactMethod{
+		Type:    d.Get("type").(string),
+		Label:   d.Get("label").(string),
+		Address: d.Get("address").(string),
+	}
+
+	if v, ok := d.GetOk("send_short_email"); ok {
+		contactMethod.SendShortEmail = v.(bool)
+	}
+
+	if v, ok := d.GetOk("country_code"); ok {
+		contactMethod.CountryCode = v.(int)
+	}
+
+	if v, ok := d.GetOk("enabled"); ok {
+		contactMethod.Enabled = v.(bool)
+	}
+
+	return contactMethod
+}
+
+func resourcePagerDutyUserContactMethodParseID(id string) (string, string) {
+	// userID, contactMethodID
+	parts := strings.Split(id, ":")
+	return parts[0], parts[1]
+}
+
+func resourcePagerDutyUserContactMethodCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*pagerduty.Client)
+
+	userID := d.Get("user_id").(string)
+
+	contactMethod := buildUserContactMethodStruct(d)
+
+	resp, _, err := client.Users.CreateContactMethod(userID, contactMethod)
+	if err != nil {
+		return err
+	}
+
+	d.SetId(fmt.Sprintf("%s:%s", userID, resp.ID))
+
+	return resourcePagerDutyUserContactMethodRead(d, meta)
+}
+
+func resourcePagerDutyUserContactMethodRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*pagerduty.Client)
+
+	userID, cmID := resourcePagerDutyUserContactMethodParseID(d.Id())
+
+	resp, _, err := client.Users.GetContactMethod(userID, cmID)
+	if err != nil {
+		return handleNotFoundError(err, d)
+	}
+
+	d.Set("address", resp.Address)
+	d.Set("blacklisted", resp.BlackListed)
+	d.Set("contact_method_id", resp.ID)
+	d.Set("country_code", resp.CountryCode)
+	d.Set("enabled", resp.Enabled)
+	d.Set("label", resp.Label)
+	d.Set("send_short_email", resp.SendShortEmail)
+	d.Set("type", resp.Type)
+	d.Set("user_id", userID)
+
+	return nil
+}
+
+func resourcePagerDutyUserContactMethodUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*pagerduty.Client)
+
+	contactMethod := buildUserContactMethodStruct(d)
+
+	log.Printf("[INFO] Updating PagerDuty user contact method %s", d.Id())
+
+	userID, cmID := resourcePagerDutyUserContactMethodParseID(d.Id())
+
+	if _, _, err := client.Users.UpdateContactMethod(userID, cmID, contactMethod); err != nil {
+		return err
+	}
+
+	return resourcePagerDutyUserContactMethodRead(d, meta)
+}
+
+func resourcePagerDutyUserContactMethodDelete(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*pagerduty.Client)
+
+	log.Printf("[INFO] Deleting PagerDuty user contact method %s", d.Id())
+
+	userID, cmID := resourcePagerDutyUserContactMethodParseID(d.Id())
+
+	if _, err := client.Users.DeleteContactMethod(userID, cmID); err != nil {
```

Comment:
If possible we should check the error here to see if it's a type `404` or otherwise "user not found" type error, and go ahead and log that the user was not found, then call `d.SetId("")`. This way, if a user contact method was destroyed externally from Terraform, we don't get stuck here unable to remove it from state because the API can't find it.

Reply:
Ah good catch :+1: changed it now so this should now be handled via the `handleNotFoundError` func :)